### PR TITLE
KAFKA-6718 / A Rack awareness for Kafka Streams

### PR DESCRIPTION
**Pull request by lkokhreidze**
[KIP-708](https://cwiki.apache.org/confluence/display/KAFKA/KIP-708%3A+Rack+awareness+for+Kafka+Streams) implementation.

Notable changes:
- Added required configuration options in `StreamsConfig`. Configurations are guarded by the validation rules and corresponding unit tests are present in `StreamsConfigTest`.
- `SubscriptionInfo` version bump to `10` and added necessary fields for encoding client tags.
- Moved standby task assignment logic behind `StandbyTaskAssignor` abstract class.
- `StandbyTaskAssignorInitializer` decides which of the `StandbyTaskAssignor` implementations to use based on `AssignmentConfigs`
- When `AssignmentConfigs#rackAwareAssignmentTags` is present, new `ClientTagAwareStandbyTaskAssignor` will be chosen which distributes the standby tasks based on client tags and configured `rackAwareAssignmentTags`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 1]
```diff
@@ -1113,6 +1162,33 @@ private void verifyEOSTransactionTimeoutCompatibility() {
             configUpdates.put(COMMIT_INTERVAL_MS_CONFIG, EOS_DEFAULT_COMMIT_INTERVAL_MS);
         }
 
+        // validate rack awareness related configuration
+        final short maxClientTagKeyValueLength = getShort(MAX_CLIENT_TAG_KEY_VALUE_LENGTH_CONFIG);
+        final List<String> rackAwareAssignmentTags = getList(RACK_AWARE_ASSIGNMENT_TAGS_CONFIG);
+        final Map<String, String> clientTags = getClientTags();
+
+        for (final String rackAwareAssignmentTag : rackAwareAssignmentTags) {
```
Comment (lkokhreidze) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1169]: Not sure if doing validations in `postProcessParsedConfig` method is okay but couldn't find any better place.

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.apache.kafka.streams.integration.utils.IntegrationTestUtils.safeUniqueTestName;
+
+@Category({IntegrationTest.class})
+public class RackAwarenessIntegrationTest {
```
Comment (lkokhreidze) [inline streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:58]: Will try to figure out more test cases, but those should already be good starting point.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignorTest.java [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+import static org.apache.kafka.streams.processor.internals.assignment.AssignmentTestUtils.TASK_0_1;
+import static org.apache.kafka.streams.processor.internals.assignment.AssignmentTestUtils.TASK_0_2;
+import static org.apache.kafka.streams.processor.internals.assignment.AssignmentTestUtils.TASK_1_0;
+import static org.apache.kafka.streams.processor.internals.assignment.AssignmentTestUtils.TASK_1_1;
+import static org.apache.kafka.streams.processor.internals.assignment.AssignmentTestUtils.TASK_1_2;
+import static org.apache.kafka.streams.processor.internals.assignment.AssignmentTestUtils.uuidForInt;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class ClientTagAwareStandbyTaskAssignorTest {
```
Comment (lkokhreidze) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignorTest.java:47]: Will try to figure out more test cases, but those should already be good starting point.

**Comment by lkokhreidze** [comment]
Call for review @vvcephei @cadonna @ableegoldman

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 4]
```diff
@@ -33,8 +31,10 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.function.BiConsumer;
+import java.util.function.Function;
```
Comment (lkokhreidze) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:34]: Sorry for the un-related changes :( happy to revert if it adds too much noise during the review.

**Comment by cadonna** [comment]
@lkokhreidze Thank you for the PR!
This is quite a large PR. Could you split it up into smaller PRs that are easier to review?
One option could be to have a separate PR for:
- standby task assignor
- config changes 
- subscription info changes
- ... 

Opening small PRs usually increases the review quality and decreases time to the first review.

**Comment by lkokhreidze** [comment]
Hi @cadonna
Thanks for the reply and suggestion. Will it be okay to have smaller PRs merged into trunk without having full functionality in place? Afair, when I worked on https://github.com/apache/kafka/pull/7170, it was one of the motivations to have full functionality in trunk in one go.
If having partial implementation in trunk is OK, I can split this PR into smaller chunks.

**Comment by cadonna** [comment]
As far as I can see from skimming the PR, you could have a PR for the standby task assignors with corresponding unit tests without changing any existing public functionality and without adding any partial public functionality. So I guess, it would be fine to have a PR for the standby task assignors. To give context to the reviewers, you should motivate the change in the PR description.

Maybe you can do similar with the subscription changes and with some other aspects of this PR. Probably, it would be good to have the config PR that makes the functionality available to the public last, so that the functionality is never partly publicly available.

I do not know the details of  #7170, but most people with which I have talked about PR sizes agree that if it is possible to split them into smaller pieces, we should do it.

Note, that the title of the PR should start with the issue ID of the JIRA tickets which in this case would be KAFKA-6718. In this way a link to the PR will show up on the ticket. You can also open multiple PRs whose title starts with KAFKA-6718. A link for each PR will then show up in the ticket.

**Comment by lkokhreidze** [comment]
Sounds good, will try to split the PR into smaller chunks.

**Comment by lkokhreidze** [comment]
Closing this one as the PR is split into multiple smaller PRs.

**Pull request closed without merging**